docker/configure_workers_and_start.py

```diff
@@ -0,0 +1,422 @@
+    # the main Synapse process as well as all workers.
+    # It is intended mainly for disabling functionality when certain workers are spun up,
+    # and adding a replication listener.
+
+    # First read the original config file and extract the listeners block. Then we'll add
+    # another listener for replication. Later we'll write out the result.
+    listeners = [
+        {
+            "port": 9093,
+            "bind_address": "127.0.0.1",
+            "type": "http",
+            "resources": [{"names": ["replication"]}],
+        }
+    ]
+    with open(config_path) as file_stream:
+        original_config = yaml.safe_load(file_stream)
+        original_listeners = original_config.get("listeners")
+        if original_listeners:
+            listeners += original_listeners
+
+    shared_config = yaml.dump({"listeners": listeners})
+
+    # Don't forget to enable redis support!
+    shared_config += """
+redis:
+    enabled: true
+"""
+
+    # The supervisord config
+    # Supervisord will be in charge of running everything, from redis to nginx to Synapse
+    # and all of its worker processes. We define a few services we'll know will run now.
+    supervisord_config = """
+[supervisord]
+nodaemon=true
+
+[program:nginx]
+command=/usr/sbin/nginx -g "daemon off;"
+priority=500
+stdout_logfile=/dev/stdout
+stdout_logfile_maxbytes=0
+stderr_logfile=/dev/stderr
+stderr_logfile_maxbytes=0
+username=www-data
+autorestart=true
+
+[program:redis]
+command=/usr/bin/redis-server /etc/redis/redis.conf --daemonize no
+priority=1
+stdout_logfile=/dev/stdout
+stdout_logfile_maxbytes=0
+stderr_logfile=/dev/stderr
+stderr_logfile_maxbytes=0
+username=redis
+autorestart=true
+
+[program:synapse_main]
+command=/usr/local/bin/python -m synapse.app.homeserver \
+    --config-path="%s" \
+    --config-path=/conf/workers/shared.yaml
+priority=10
+# Log startup failures to supervisord's stdout/err
+# Regular synapse logs will still go in the configured data directory
+stdout_logfile=/dev/stdout
+stdout_logfile_maxbytes=0
+stderr_logfile=/dev/stderr
+stderr_logfile_maxbytes=0
+autorestart=unexpected
+exitcodes=0
+
+""" % (
+        config_path,
+    )
+
+    # An nginx site config that will be amended to depending on the workers that are
+    # spun up. To be placed in /etc/nginx/conf.d
+    nginx_config_template_header = """
+server {
+    # Listen on an unoccupied port number
+    listen 8008;
+    listen [::]:8008;
+
+    server_name localhost;
+
+    # Nginx by default only allows file uploads up to 1M in size
+    # Increase client_max_body_size to match max_upload_size defined in homeserver.yaml
+    client_max_body_size 100M;
+    """
+    nginx_config_body = ""  # to modify below
+    nginx_config_template_end = """
+    # Send all other traffic to the main process
+    location ~* ^(\\/_matrix|\\/_synapse) {
+        proxy_pass http://localhost:8080;
+        proxy_set_header X-Forwarded-For $remote_addr;
+    }
+}
```

Comment:
Why nginx? We use haproxy everywhere else.

Reply:
Nginx was the solution that was reached for first, back in the [design doc](https://docs.google.com/document/d/1XcwrO_JX2erurW82czk7ivsfQUQSFMsrGPmPfKqnGnY/edit#), mostly due to its simple config format. I can replace it with HAProxy instead, though it'd be a bit of a pain.

Reply:
Although I will note to myself that there is quite a lot we can crib from https://github.com/matrix-org/sytest/blob/c75ff0449b5b62bdebfdb9024170988b519628d9/lib/SyTest/Homeserver/Synapse.pm#L1145-L1305.

Reply:
It looks like if we want to support more than just the [basic load-balancing techniques](https://nginx.org/en/docs/http/load_balancing.html), then we'll likely need to switch over to HAProxy.

I'll check how big of a task this will be, but at the moment I'm leaning towards just supporting round-robin LB for now with nginx, and then switch to HAProxy after the fact. Doing so shouldn't require any changes in Synapse/Complement.

Reply:
So nginx does actually support load-balancing by request IP, or by hashing something in the request: https://nginx.org/en/docs/http/ngx_http_upstream_module.html#hash

But actually thinking about it again I don't think we really want any load-balancing techniques other than round-robin for the Complement usecase. We certainly don't want to load-balance by IP, as all requests from our test infra would be coming from localhost.

In fact I think what we really want is consistency in load-balancing between test runs. And round-robin gives us exactly that.

Reply:
The main reason to use haproxy rather than nginx is that haproxy is kinda what we "officially" support. It doesn't matter too much but its nice to be consitent